contracts/pol/CarbonPOL.sol

```diff
@@ -0,0 +1,303 @@
+        return _marketPriceMultiply;
+    }
+
+    /**
+     * @inheritdoc ICarbonPOL
+     */
+    function priceDecayHalfLife() external view returns (uint32) {
+        return _priceDecayHalfLife;
+    }
+
+    /**
+     * @inheritdoc ICarbonPOL
+     */
+    function tradingEnabled(Token token) external view returns (bool) {
+        return _tradingEnabled(token);
+    }
+
+    /**
+     * @inheritdoc ICarbonPOL
+     */
+    function expectedTradeReturn(Token token, uint128 ethAmount) external view returns (uint128 tokenAmount) {
+        // return 0 if trading is not enabled for token
+        if (!_tradingEnabled(token)) {
+            return 0;
+        }
+        Price memory currentPrice = tokenPrice(token);
+        // revert if ethAmount in price is equal to 0
+        if (currentPrice.ethAmount == 0) {
+            revert InvalidPrice();
+        }
+        // multiply the token amount by the eth amount / total eth amount ratio to get the actual tokens received
+        tokenAmount = uint128(MathEx.mulDivF(currentPrice.tokenAmount, ethAmount, currentPrice.ethAmount));
+        // revert if not enough token balance
+        if (tokenAmount > uint128(token.balanceOf(address(this)))) {
+            revert InsufficientTokenBalance();
+        }
+    }
+
+    /**
+     * @inheritdoc ICarbonPOL
+     */
+    function expectedTradeInput(Token token, uint128 tokenAmount) public view returns (uint128 ethAmount) {
+        // return 0 if trading is not enabled for token
+        if (!_tradingEnabled(token)) {
+            return 0;
+        }
+        // revert if not enough token balance for trade
+        if (tokenAmount > token.balanceOf(address(this))) {
+            revert InsufficientTokenBalance();
+        }
+        Price memory currentPrice = tokenPrice(token);
+        // multiply the eth amount by the token amount / total token amount ratio to get the actual eth to send
+        return uint128(MathEx.mulDivF(currentPrice.ethAmount, tokenAmount, currentPrice.tokenAmount));
+    }
+
+    /**
+     * @inheritdoc ICarbonPOL
+     */
+    function tokenPrice(Token token) public view returns (Price memory price) {
+        // return 0 if trading hasn't been enabled for a token
+        if (!_tradingEnabled(token)) {
+            return Price({ tokenAmount: 0, ethAmount: 0 });
+        }
+        // get initial price as set by enableTrading
+        price = _initialPrice[token];
+        // calculate the actual price by multiplying the eth amount by the factor
+        price.ethAmount *= _marketPriceMultiply;
+        // get time elapsed since trading was enabled
+        uint32 timeElapsed = uint32(block.timestamp) - _tradingStartTimes[token];
+        // get the current price by adjusting the eth amount with the exp decay formula
+        price.ethAmount = uint128(ExpDecayMath.calcExpDecay(price.ethAmount, timeElapsed, _priceDecayHalfLife));
+        // return the price
+        return price;
```

Comment:
This solution can be greatly simplified, as well as improved for accuracy.

Function `calcExpDecay` performs a division, which is redundant in this case, since the `Price` object already represents `ethAmount / tokenAmount`, so there is no need for an additional division.

So in short, you can just use whatever function `exp2` returns, which means that you don't even need to include function `calcExpDecay` as part of the project, and you don't need to worry about any additional testing of your changes in that function (since in contrast to it, function `exp2` remains unchanged).

And as an additional bonus, you avoid the loss of accuracy which occurs when function `calcExpDecay` calls function `mulDivF`.

In short, here is the code:
```
function tokenPrice(Token token) public view returns (Price memory) {
    Price memory initialPrice = _initialPrice[token];

    Fraction memory timeRatio = Fraction({
        n: block.timestamp - _tradingStartTimes[token],
        d: _priceDecayHalfLife }
    );

    Fraction memory decayFactor = MathEx.exp2(timeRatio);

    return Price({
        ethAmount: initialPrice.ethAmount * decayFactor.d * _marketPriceMultiply,
        tokenAmount: initialPrice.tokenAmount * decayFactor.n
    });
}
```

Lastly, since we already have a `Fraction` object, you might want to consider it instead of the `Price` object (though the `Fraction` object is less verbose than the `Price` object, so just FYI to consider).

Reply:
Prone to overflow - decayFactor.d is actually 2 ^ _priceDecayHalfLife.
So for any _priceDecayHalfLife > 256 and initialPrice.ethAmount >= 1 the value overflows.
Same goes for tokenAmount.

Reply:
We're not calculating `2 ^ halfLife`, we're calculating `2 ^ timeElapsed / halfLife`.

And the value of `timeElapsed / halfLife` cannot exceed ~23 (16 / ln 2) regardless, due to a technical limitation in function `exp2`.

If anything, the overflow here would occur because each component in the `Fraction` returned by `exp2` is scaled up by `2 ^ 127`.

And you can handle that by declaring the `Price` components as `uint256` instead of `uint128` (in case they aren't already declared as such).

Reply:
I tried this approach and it reverted with overflow.
I already implemented the other approach and wrote the tests for it.

Reply:
Logic here is a bit confusing and will need further testing to confirm and verify it's working correctly.
Sticking for my current solution for now, waiting for further input.
Your original suggestion is prone to overflow if using uint128, and switching to uint256 means we disregard the gas savings. I don't see much benefit of implementing it at this stage.